kernel/comps/virtio/src/device/network/device.rs

```diff
@@ -139,22 +160,29 @@ impl NetworkDevice {
         Ok(())
     }
 
-    /// Add a rx buffer to recv queue
-    /// FIEME: Replace rx_buffer with VM segment-based data structure to use dma mapping.
+    /// Adds a `RxBuffer` to the receive queue.
     fn add_rx_buffer(&mut self, rx_buffer: RxBuffer) -> Result<(), VirtioNetError> {
         let token = self
             .recv_queue
             .add_dma_buf(&[], &[&rx_buffer])
             .map_err(queue_to_network_error)?;
         assert!(self.rx_buffers.put_at(token as usize, rx_buffer).is_none());
-        if self.recv_queue.should_notify() {
-            self.recv_queue.notify();
+
+        self.poll_stat.received_packet += 1;
+
+        if self.poll_stat.received_packet == QUEUE_SIZE as _ {
+            // If we know there are no free buffers for receiving,
+            // we will notify the receive queue as soon as possible
```

Comment:
This should be `debug!("notify send queue: sent {} packets")` to make the log message reliable (otherwise you have to guess if this path is used to judge if the log messages really contain all the packets sent).

Maybe this also means that we want a method to do the notification and statistics reset?
```rust
            debug!(
                "notify receive queue: received {} packets",
                self.poll_stat.received_packet
            );
            if self.recv_queue.should_notify() {
                self.recv_queue.notify();
            }
            self.poll_stat.received_packet = 0;
```